### PR TITLE
Yaml map sequences

### PR DESCRIPTION
**Pull request by matthid**
This is basically the implementation of https://github.com/fsprojects/FSharp.Configuration/issues/51

But with some major limitations:
 * We only use the first list item for the provided type
 * If you use different items it will probably break one way or another

Some notes:
 * I created `static member CreateResizeArray<'a>(data : 'a seq) :ResizeArray<'a> = ResizeArray<'a>(data)` because I did not find a way/api to call a constructor of a generic type with a provided type parameter.
 * I'm not sure why we sort on update? Why would differently ordered lists not be different?

Please be kind as these are my first experiments with type providers (a painful experience) :)

### CODE REVIEW

**Comment by vasily-kirichenko** [comment]
It's great! Thanks.
Could you fix AppVeyor build?

File: src/FSharp.Configuration/ProvidedTypes.fs [hunk 1]
```diff
@@ -941,7 +942,9 @@ type ProvidedSymbolType(kind: SymbolKind, args: Type list) =
         | SymbolKind.Array _ -> typeof<System.Array>
         | SymbolKind.Pointer -> typeof<System.ValueType>
         | SymbolKind.ByRef -> typeof<System.ValueType>
-        | SymbolKind.Generic gty  -> ProvidedSymbolType.convType args gty.BaseType
+        | SymbolKind.Generic gty  -> 
+            if gty.BaseType = null then null else
+            ProvidedSymbolType.convType args gty.BaseType
```
Comment (vasily-kirichenko) [inline src/FSharp.Configuration/ProvidedTypes.fs]: We should not change `ProvidedTypes.fs/fsi`. We copy them from FSharp.TypeProviders.StarterPack repo. If you find a bug, please make a PR to that repo as well.

Comment (matthid) [inline src/FSharp.Configuration/ProvidedTypes.fs]: Sorry that was not obvious for me. Is there a reason to not reference the corresponding nuget package in this case? Can you also briefly explain (or send me a link) why we even need to copy them?

**Comment by vasily-kirichenko** [comment]
image

**Comment by vasily-kirichenko** [comment]
image

**Comment by matthid** [comment]
Sure I will try to fix appveyor (it took quite some time to start yesterday). I'm not even sure why it would break... I didn't even change anything there :/.

Second picture:
Some of the YamlExceptions come from "Settings.yaml" and where there even before I started (no idea why). Sometimes SharpYaml seems to go crasy and you need to restart/rebuild.

First picture:
This is obviously working for me, otherwise I wouldn't have sent this PR :(

**Comment by matthid** [comment]
It seems to work in AppVeyor as well now.

**Comment by vasily-kirichenko** [comment]
Aha, if I change `Lists.yaml` as following:

```yaml
items:
    - part_no:   A4786
      descrip:   Water Bucket (Filled)
      price:     1,47
      quantity:  4

    - part_no:   E1628
      descrip:   High Heeled "Ruby" Slippers
      size:      8
      price:     100,27
      quantity:  1
```
The exception has gone. 
So, we should be able to parse floats with different decimal separators. I'm afraid it's a SharpYaml bug.

**Comment by vasily-kirichenko** [comment]
For now, could change type of `price` to int? I mean `price: 147`, `price: 10027`.

**Comment by matthid** [comment]
Sure, But the exception is catched somewhere and it seems to work anyway, right? I'm a bit confused now as the tests seem to run fine.

**Comment by matthid** [comment]
Ah so you saying its not working on your machine because of localization, is that correct?

File: src/FSharp.Configuration/ProvidedTypes.fs (hunk 1 above)
Comment (vasily-kirichenko) [inline src/FSharp.Configuration/ProvidedTypes.fs]: This is just as it's intended to be used from the very beginning :) Every type provider must have it's own copy of these two files. The single central place where they should be changed is FSharp.TypeProviders.StarterPack. If you need deeper explanation, ask @dsyme.

**Comment by vasily-kirichenko** [comment]
Yes.

File: tests/FSharp.Configuration.Tests/YamlProvider.Tests.fs [hunk 2]
```diff
@@ -218,4 +218,59 @@ Mail:
...
+    settings.items.[0].descrip |> should equal "Some description"
+    settings.items.[0].quantity |> should equal 14
+    
+    settings.items.[2].part_no |> should equal "E1628"
+    settings.items.[2].descrip |> should equal "High Heeled \"Ruby\" Slippers"
+    settings.items.[2].quantity |> should equal 1
+
+[<Ignore>]
+[<Test>]
+let ``Check that list defaults are OK``() =
```
Comment (vasily-kirichenko) [inline tests/FSharp.Configuration.Tests/YamlProvider.Tests.fs]: Could we remove this test?

**Comment by matthid** [comment]
Thanks for figuring that out. Ok that should now be fixed as well (waiting for AppVeyor). Can you open a issue over there at SharpYaml?

File: tests/FSharp.Configuration.Tests/YamlProvider.Tests.fs (hunk 2 above)
Comment (matthid) [inline tests/FSharp.Configuration.Tests/YamlProvider.Tests.fs]: Sure, but it should work eventually (if somebody implements the proper default behavior). Do you not want to keep them in the mean time?

Comment (vasily-kirichenko) [inline tests/FSharp.Configuration.Tests/YamlProvider.Tests.fs]: Ah, ok.

**Comment by matthid** [comment]
Thanks for merging. I sent the changes to https://github.com/fsprojects/FSharp.TypeProviders.StarterPack/pull/54 as well. I noticed that the files have already diverged, so maybe we should add the nuget package or use paket to reference their files (See https://github.com/fsprojects/FSharp.TypeProviders.StarterPack/issues/44)? It would be also more clear for future collaborators to not touch those files.

File: src/FSharp.Configuration/ProvidedTypes.fs (hunk 1 above)
Comment (matthid) [inline src/FSharp.Configuration/ProvidedTypes.fs]: Ok thanks, very weird :) .

**Comment by mavnn** [comment]
A paket file reference would be the preferred method at the moment, yes.

**Pull request merged**